### PR TITLE
Implement constant buffers for HDRP global variables.

### PR DESCRIPTION
**Pull request by JulienIgnace-Unity**
### Purpose of this PR
Currently in HDRP, all global shader variables are set through the SetGlobalXXX APIs.
This is a problem for several reasons:
- Code readability: Setup of globals is spread everywhere around the code, which makes it very difficult to track things down.
- Performance: We do hundreds of calls to SetGlobalXXX APIs which goes through a C#/C++ transition everytime which is slow on the main thread. This is also slower on the render thread side because constant buffers have to be reconstructed and bound internally very often because of that.

In order to fix this, this PR will convert all HDRP global shader variables to use the new explicit Constant Buffer API.
Content of the PR:
- Added constant buffer handling to HLSL generation.
- Added a new ConstantBuffer class to abstract the API.
- Converted constant buffers are now defined in C# and generate the proper HLSL structure (this allows us to keep consistency between the two sides)
- Centralized update of those constant buffers in specific functions in HDRenderPipeline.cs and the various contributor to global variables (hdCamera, various lighting systems, etc.)

@stramit @phi-lira added you to take a look at the proposed API for ConstantBuffer abstraction.

---
### Testing status

**Manual Tests**: What did you do?
- [x] Opened test project + Run graphic tests locally
- [x] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo + Prefab overrides + Alignment in Preset
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in developer mode, you have a button at end of resources that check the paths)

@Unity-Technologies/gfx-qa-hdrp this PR changes a LOT of core code in HDRP so testing this specifically is complicated. I'd say an overall smoke test combined with automated graphics tests should do it.

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs [hunk 1]
```diff
@@ -980,62 +980,8 @@ void RegisterLightingDebug()
...
-                                                id = HDShaderIDs._DepthPyramidScale;
-                                                break;
-                                        }
-                                        var size = Shader.GetGlobalVector(id);
-                                        float lodCount = size.z;
-                                        return (uint)lodCount;
-                                    }
-                            }
+
+                            new DebugUI.FloatField { displayName = "Fullscreen Debug Mip", getter = () => data.fullscreenDebugMip, setter = value => data.fullscreenDebugMip = value, min = () => 0f, max = () => 1f, incStep = 0.05f }
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs:984]: Why this change here? or was it dead code?

Comment (JulienIgnace-Unity) [reply com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs:984]: Because the old code relied on doing things like Shader.GetGlobalVector with the old _ColorPyramidScale parameter. This was already bad at the time and it cannot work with constant buffers. Besides, this was done only to get the lod count and was not actually used because the debug item is a slider between 0 and 1 anyway.

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/ShaderVariablesLightLoop.cs.hlsl [hunk 2]
```diff
@@ -1,49 +0,0 @@
...
-    float4 _CookieAtlasSize;
-    float4 _CookieAtlasData;
-    float4 _PlanarAtlasData;
-    float _MicroShadowOpacity;
-    float _DirectionalTransmissionMultiplier;
-    uint _NumTileFtplX;
-    uint _NumTileFtplY;
-    float g_fClustScale;
-    float g_fClustBase;
-    float g_fNearPlane;
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/ShaderVariablesLightLoop.cs.hlsl:37]: could you check that you have correctly remove all meta files as well?

**Comment by sebastienlagarde** [comment]
Issue for async compute in player will be fix in a following PR. Merging

@JulienIgnace-Unity think about adding an upgrade guide doc about it
